test/randtilegen.c

```diff
@@ -0,0 +1,264 @@
+					    tileData[index][y ^ ((rotation & 2) ? 7 : 0)][x ^ ((rotation & 1) ? 7 : 0)];
+				}
+			}
+		} else {
+			switch (attributes[p].nbColors) {
+			case 2:
+				for (uint8_t y = 0; y < 8; y++)
+					for (uint8_t x = 0; x < 8; x++)
+						tileData[p][y][x] = getRandomBits(1);
+				break;
+			case 4:
+				for (uint8_t y = 0; y < 8; y++)
+					for (uint8_t x = 0; x < 8; x++)
+						tileData[p][y][x] = getRandomBits(2);
+				break;
+			default:
+				for (uint8_t y = 0; y < 8; y++)
+					for (uint8_t x = 0; x < 8; x++) {
+						do {
+							index = getRandomBits(2);
+						} while (index == 3);
+						tileData[p][y][x] = index;
+					}
+			}
+		}
+	}
+
+	uint16_t colors[10];
+	for (uint8_t p = 0; p < 10; p++) {
+		colors[p] = getRandomBits(15);
+	}
+	// Randomly make color #0 of all palettes transparent
+	if (!getRandomBits(2)) {
+		colors[0] |= 0x8000;
+		colors[5] |= 0x8000;
+	}
+
+	uint16_t palettes[60][4];
+	for (uint8_t p = 0; p < 60; p++) {
+		const uint16_t *subpal = colors;
+		if (p & 1) {
+			subpal += 5;
+		}
+		uint8_t total = 0;
+		for (uint8_t index = 0; index < 5; index++) {
+			if (p & (2 << index)) {
+				palettes[p][total++] = subpal[index];
+			}
+		}
+	}
+
+	png_set_IHDR(png, pngInfo, width * 8, height * 8, 8, PNG_COLOR_TYPE_RGB_ALPHA,
+	             getRandomBits(1) ? PNG_INTERLACE_NONE : PNG_INTERLACE_ADAM7,
+	             PNG_COMPRESSION_TYPE_DEFAULT, PNG_FILTER_TYPE_DEFAULT);
+
+	// While it would be nice to write the image little by little, I really don't want to handle
+	// interlacing myself. (We're doing interlacing to test that RGBGFX correctly handles it.)
+	uint8_t const SIZEOF_PIXEL = 4; // Each pixel is 4 bytes (RGBA @ 8 bits/component)
+	uint8_t data[height * 8 * width * 8 * SIZEOF_PIXEL];
+	uint8_t *rowPtrs[height * 8];
+	for (uint8_t y = 0; y < height * 8; ++y) {
+		rowPtrs[y] = &data[y * width * 8 * SIZEOF_PIXEL];
+	}
+
+	for (uint8_t p = 0; p < nbTiles; p++) {
+		uint8_t tx = 8 * (p % width), ty = 8 * (p / width);
+		for (uint8_t y = 0; y < 8; y++) {
+			uint8_t * const row = rowPtrs[ty + y];
+			for (uint8_t x = 0; x < 8; x++) {
+				uint8_t * const pixel = &row[(tx + x) * SIZEOF_PIXEL];
+				uint16_t color = palettes[attributes[p].palette][tileData[p][y][x]];
+				pixel[0] = _5to8(color & 0x1F);
+				pixel[1] = _5to8(color >> 5 & 0x1F);
+				pixel[2] = _5to8(color >> 10 & 0x1F);
+				pixel[3] = color & 0x8000 ? 0x00 : 0xFF;
+			}
+		}
+	}
+	png_set_rows(png, pngInfo, rowPtrs);
+	png_write_png(png, pngInfo, PNG_TRANSFORM_IDENTITY, NULL);
+}
+
+int main(int argc, char **argv) {
+	if (argc < 2) {
+		fputs("usage: randtilegen <basename> [<basename> [...]]\n", stderr);
+		return 2;
+	}
+
+	size_t maxBasenameLen = 0;
+	for (int index = 1; index < argc; index++) {
+		size_t length = strlen(argv[index]);
+		if (length > maxBasenameLen) {
+			maxBasenameLen = length;
+		}
+	}
+
+	char filename[maxBasenameLen + sizeof("65535.png")];
+	for (uint16_t i = 0;; i++) { // 65k images ought to be enough
```

Comment:
Instead of relying on the maximum size of `uint16_t` here, and needing an explicit `if (i == UINT16_MAX) { break; }` at the end anyway, I'd do `for (unsigned int i = 0; i < MAX_IMAGES; i++)` and use whatever max you prefer (65536, 10000, etc). Or even use a signed `int i` so overflow will be undefined and the compiler will know the loop executes exactly `MAX_IMAGES` times.

Reply:
The "65k" comment is only to say that putting a hard limit of 65k should still be sufficient for most use cases; the `break` is only a fail-safe to avoid generating an infinite amount of them if e.g. piping in `/dev/urandom`. The program exits when the input stream is exhausted, so not necessarily after `MAX_IMAGES` images.